MINOR: Set `group.coordinator.rebalance.protocols` to `classic,consumer` by default

dajac opened the pull request:
This patch changes the default configuration of `group.coordinator.rebalance.protocols` to `classic,consumer`. It also updates various tests that were specifying the new default value.


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
MINOR: Set `group.coordinator.rebalance.protocols` to `classic,consumer` by default

dajac commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -576,19 +576,13 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
       throw new ConfigException(s"Disabling the '${GroupType.CLASSIC}' protocol is not supported.")
     }
     if (protocols.contains(GroupType.CONSUMER)) {
-      if (processRoles.isEmpty) {
```
I had to soften the validation here because ZK also has the new default value now. Keep in mind that ZK will be remove entirely in 4.0 so it won't be an issue for the next release.

chia7712 commented on core/src/test/java/kafka/test/ClusterInstance.java:
```diff
@@ -160,15 +159,11 @@ default Admin createAdminClient() {
-
-        if (serverProperties.getOrDefault(GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, "").contains("consumer")) {
-            supportedGroupProtocols.add(CONSUMER);
+        if (isKRaftTest()) {
```
The `CONSUMER` protocol can be disabled even though it is kraft mode, right? Maybe we can check the `KafkaAPIs` directly. for example:
```java
    default Set<GroupProtocol> supportedGroupProtocols() {
        if (brokers().values().stream().allMatch(b -> b.dataPlaneRequestProcessor().isConsumerGroupProtocolEnabled())) {
            return mkSet(CLASSIC, CONSUMER);
        } else {
            return Collections.singleton(CLASSIC);
        }
    }
```

dajac replied:
Good idea. Let me try this.

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java:
```diff
@@ -76,9 +76,9 @@ public class GroupCoordinatorConfig {
             "The " + Group.GroupType.SHARE + " rebalance protocol is in early access and therefore must not be used in production.";
-    public static final List<String> GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DEFAULT = Collections.singletonList(Group.GroupType.CLASSIC.toString());
+    public static final List<String> GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DEFAULT =
+        Arrays.asList(Group.GroupType.CLASSIC.toString(), Group.GroupType.CONSUMER.toString());
```
Please make it immutable.

chia7712 commented on core/src/test/java/kafka/test/ClusterTestExtensionsTest.java:
```diff
@@ -191,45 +189,17 @@ public void testDefaults(ClusterInstance clusterInstance) {
-        @ClusterTest(types = {Type.ZK, Type.KRAFT, Type.CO_KRAFT}, serverProperties = {
-            @ClusterConfigProperty(key = NEW_GROUP_COORDINATOR_ENABLE_CONFIG, value = "false"),
-        }),
-        @ClusterTest(types = {Type.ZK, Type.KRAFT, Type.CO_KRAFT}, serverProperties = {
```
why remove this test case? it seems to be legal configs

dajac replied:
It did not work with my implementation but it would work with yours. Let me bring it back.

chia7712 commented on core/src/test/java/kafka/test/ClusterTestExtensionsTest.java:
```diff
@@ -191,45 +189,17 @@ public void testDefaults(ClusterInstance clusterInstance) {
-            @ClusterConfigProperty(key = NEW_GROUP_COORDINATOR_ENABLE_CONFIG, value = "true"),
-            @ClusterConfigProperty(key = GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, value = "classic"),
-        }),
-        @ClusterTest(types = {Type.ZK, Type.KRAFT, Type.CO_KRAFT}, serverProperties = {
```
ditto

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandTestUtils.java:
```diff
@@ -79,34 +77,34 @@ private ConsumerGroupCommandTestUtils() {
         Map<String, String> serverProperties = new HashMap<>();
         serverProperties.put(OFFSETS_TOPIC_PARTITIONS_CONFIG, "1");
         serverProperties.put(OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, "1");
         serverProperties.put(NEW_GROUP_COORDINATOR_ENABLE_CONFIG, "true");
```
this can be removed too.

dajac replied:
I was debating whether I should remove those in this PR. I can do it.

chia7712 commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -576,19 +576,13 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
-      if (!isNewGroupCoordinatorEnabled) {
-        throw new ConfigException(s"The new '${GroupType.CONSUMER}' rebalance protocol is only supported by the new group coordinator.")
+      if (processRoles.isEmpty || !isNewGroupCoordinatorEnabled) {
+        warn(s"The new '${GroupType.CONSUMER}' rebalance protocol is only supported in KRaft cluster with the new group coordinator.")
```
Maybe `KafkaAPIs#isConsumerGroupProtocolEnabled` needs similar log when the `isConsumerRebalanceProtocolSupported` return false and `CONSUMER` protocol is enabled.

https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaApis.scala#L3816

dajac replied:
Hum... This is too much in my opinion because it would log on every calls from the consumers. Having a warning at the beginning seems enough for me.

chia7712 replied:
> it would log on every calls from the consumers

you are right :100:

chia7712 reviewed (commented):
@dajac thanks for this patch. 

BTW, should we remove the redundant config `key = GroupCoordinatorConfig.NEW_GROUP_COORDINATOR_ENABLE_CONFIG, value = "true"`?

dajac committed:
address minor comments

dajac committed:
cleanups

dajac committed:
more cleanups

dajac committed:
fix checkstyle

dajac commented:
@chia7712 Thanks for your comments. I have addressed all of them. I have also cleaned up a few more things in the integration tests. Please take a look when you get a chance.

chia7712 commented:
@dajac Could you please remove `kip848` from `PlaintextConsumerAssignorsTest`?

https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/api/PlaintextConsumerAssignorsTest.scala#L241
https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/api/PlaintextConsumerAssignorsTest.scala#L271

dajac committed:
address comment + fix tests

dajac commented:
@chia7712 I have addressed your last comment. Please take a look when you get a chance.

dajac committed:
Merge remote-tracking branch 'upstream/trunk' into djacot-minor-rebalance-config-4

dajac committed:
Merge pull request #90 from confluentinc/djacot-minor-rebalance-config-4

chia7712 reviewed (approved):
LGTM

chia7712 commented:
the failed test is traced by https://issues.apache.org/jira/browse/KAFKA-17486

I will merge it
